crates/shared/src/event_handling.rs

```diff
@@ -69,48 +72,92 @@ where
         block_retriever: B,
         contract: C,
         store: S,
-        start_sync_at_block: Option<u64>,
+        start_sync_at_block: Option<BlockNumberHash>,
```

Comment:
I just noticed, but we added a bunch of `web3` calls to get the block hash at a specific depth in various places (like the Subgraph initializers, contracts, etc).

I think it would make more sense to have:
```rust
    pub fn new(
        block_retriever: B,
        contract: C,
        store: S,
        start_sync_at_block: Option<BlockNumberHash>,
    ) -> Self { ... }
    pub async fn at_block_number(
        block_retriever: B,
        contract: C,
        store: S,
        start_sync_at_block_number: Option<u64>,
    ) -> Result<Self> {
        let start_sync_at_block = match start_sync_at_block_number {
            Some(number) => Some((number, block_retreiver.block_hash_for_number(number).await?)),
            None => None,
        };
        Ok(Self::new(block_retriever, contract, store, start_sync_at_block))
    }
```

Reply:
I was thinking of this however would need to change a bunch of constructors to async, will check again...